### DISCUSSION TITLE
Idea - Single player "skirmish" game type vs AI players

### DISCUSSION THREAD

**Discussion by ageekhere** [CorsixTH/CorsixTH#2431] [Research Room]
Before multiplayer is made the idea is to first create a "skirmish" game type with AI players (the same game type that would be with human players). So the idea is to take the multiplayer game play from theme hospital but create AI players which the human can play with. There would be a few things that would need to be set up.
1. The ability to create Ai logic in LUA, this would be things like make X room X size and set item locations within room, hire staff etc.
2. The player hospitals, which would be a mirror layout for all players
3. 1 pool of patient which all players fight over

post your ideas on this below.

**Comment by Alberth289346** [comment]
Sorry to burst your bubble, but ideas go around everywhere. Literally everybody has them, and everybody has way more ideas than they can realize in their life.

This also means you'll find it hard to get anyone to work for free on your ideas. Given a choice between work for free on own ideas versus work for free on ideas of someone else, pretty much everybody picks option 1.

As a result, if you want this to go anywhere beyond writing a page of text that nobody will pick up, start implementing it yourself.

**Reply by ageekhere** [reply]
I will give it a go, going to first wright up what would be needed in order to code an AI and go from there.
After i create the AI reference i will create a work in progress pull request.

**Comment by ageekhere** [comment]
So the first draft is out https://github.com/CorsixTH/CorsixTH/pull/2433

The first thing that i need some help is how to get and set data for an AI player from the game.
Lets use getRest() and setRest() in the ai.lua
How can i get the current AI player's rest policy and set it to a new value?

**Comment by Alberth289346** [comment]
> The first thing that i need some help is how to get and set data for an AI player from the game.

`hospital.lua` contains pretty much everything that exists for the hospitals. So all information you want to change exists there.
There is however no clean separated interface though, that should probably be created by cleaning up the code.

> How can i get the current AI player's rest policy and set it to a new value?

There is no AI player at all, there is only an AI hospital, and it does a lot of nothing at all currently.

**Comment by Alberth289346** [comment]
What about finding or creating these 2 functions as a first step?  (Or even just the "get" side, I have no idea how complicated adding the interface will be.).

**Comment by Alberth289346** [comment]
While you want the AI functions eventually, doing them all before you have an AI hospital running may be very boring. You could also make steps in the AI itself.perhaps?  
Consider an AI hospital being the same as an AI player to make life simpler, and create in some very limited form an AI hospital. that you can see in the game.

AI hospitals are created, I don't know if you get a call to do processing for making decisions (although I wouldn't be surprised if that currently works). I also don't know whether they get a parcel assigned to them.

**Reply by ageekhere** [reply]
> While you want the AI functions eventually, doing them all before you have an AI hospital running may be very boring. You could also make steps in the AI itself.perhaps?

I wanted to first get a good idea of what would be needed for an ai creator and what data that i need to get from the game for the AI.

> Consider an AI hospital being the same as an AI player to make life simpler, and create in some very limited form an AI hospital. that you can see in the game.

This is exactly the next step and during this processes i start to add the core AI functionality to ai.lua while making a simple AI.

> AI hospitals are created, I don't know if you get a call to do processing for making decisions (although I wouldn't be surprised if that currently works). I also don't know whether they get a parcel assigned to them.

The first goal is to find a way to set up an AI player as it was a human player with their own hospital where i can view their game play.
Then i can slowly add AI functionality to ai.lua while creating a basic ai to start with.

Though I will need help with setting up a hospital so it can be played as an AI like a human would.

**Comment by Alberth289346** [comment]
The hospital exists, `world.lua` makes 4 hospital instances. One for the player, and 3 that are currently not used.
Extend that by adding an `AIPlayer` instance or so to them.
I don't know if you get tick calls to do processing both for the hospital and the player, but that should be a next thing to solve.

> Though I will need help with setting up a hospital so it can be played as an AI like a human would.

I don't know what you have in mind here. You can ask concrete questions about the existing code (probably the chat is quicker than in a discussion), so you get understanding why code is as it is, but please don't expect anyone to write your code.

It will take time to understand the code base and make sense of it, spend time getting acquainted with it. Read how things are done and eventually you'll be able to find your way in it.
